service/src/main/python/jcarbon/tensorflow/callbacks.py

```diff
@@ -185,3 +188,100 @@ def on_epoch_end(self, epoch, logs=None):
         self.client.stop(self.pid)
         self.client.dump(
             self.pid, f'{self.output_path}/report-{epoch}.pb', self.signals)
+
+class JCarbonNvmlCallback(JCarbonCallback):
+    def __init__(
+            self,
+            addr='localhost:8980',
+            period_ms=DEFAULT_PERIOD_MS,
+            signals=DEFAULT_SIGNALS,
+            chunking_period_sec=DEFAULT_PERIOD_SECS):
+        super().__init__(addr, period_ms, signals)
+        self.reports = {}
+        self.sampler = NvmlSampler()
+    
+    def on_epoch_begin(self, epoch, logs = None):
+        self.last_report = None
+        self.sampler.sample()
+    
+    def on_epoch_end(self, epoch, logs = None):
+        self.sampler.sample()
+        if self.last_report is None:
+            self.last_report = [create_report(self.sampler.samples)]
+            print('in none)')
+            print(self.last_report)
+        else:
+            self.last_report.append(create_report(self.sampler.samples))
+            print(self.last_report)
+
+        self.reports[epoch] = pd.concat(list(map(
+            to_dataframe,
+            self.last_report
+        ))).to_frame().assign(
+            epoch=epoch).set_index('epoch', append=True)
+
+class JCarbonChunkingCallback3(JCarbonCallback):
+    def __init__(
+            self,
+            addr='localhost:8980',
+            period_ms=DEFAULT_PERIOD_MS,
+            signals=DEFAULT_SIGNALS,
+            chunking_period_sec=DEFAULT_PERIOD_SECS):
+        super().__init__(addr, period_ms, signals)
+        self.reports = {}
+        self.timestamps = {}
+        self.batch_timestamps = None
+        self.batch_idx = None
+        self.epoch_idx = None
+        self.chunking_period_sec = chunking_period_sec
+
+    def on_epoch_begin(self, epoch, logs=None):
+        self.time = time.time()
+        self.last_report = None
+        self.curr_batch_timestamps = None
+        if self.epoch_idx is None:
+            self.epoch_idx = 1
+        else:
+            self.epoch_idx+=1
+        self.start_jcarbon()
+    
+    def on_train_batch_begin(self, epoch, logs=None):
```

Comment:
I do not think your logic actually works. You never reset `batch_idx`. When we go from epoch 0 to epoch 1, this will still be the batch number, so it will just continue incrementing.